### PR TITLE
Add support for entity category and state class sensors

### PR DESCRIPTION
**Pull request by dshokouhi**
## Summary

* Added support for `entity_category` and `state_class` on a per entity basis
* Upon first install of this version we will register all sensors. Any time the app is updated we will register all sensors. Every 4 hours we will check and store the HA version during a sensor update and if it has changed we will register sensors.

As of now we are waiting on `mobile_app` to be updated with support allowing for sensors to be re-registered, until then this PR only works on sensors that were never registered and only when the user is on HA core 2021.11.x

~~We may want to wait for this PR to be merged first?~~ 
https://github.com/home-assistant/core/pull/58378

Chosen categories and/or state class:
|Sensor| State Class | Category|
|---|---|---|
|activity | none| |
|sleep segment| |
|sleep confidence | measurement|
|high accuracy mode|| config|
|current version ||diagnostic|
|app rx gb |total_increasing| diagnostic|
|app tx gb |total_increasing |diagnostic|
|app memroy | measurement| diagnostic
|app inactive || diagnostic
|standby bucket || diagnostic
|app importance | |diagnostic
|ringer mode ||config
|audio state | |none
|headphone | |none
|mic muted | |none
|music active | |none
|speakerphone | |none
|volume alarm, call, music, notification || config
|battery level, battery temperature | measurement |diagnostic
|battery state, is charging, charger type, battery health ||diagnostic
|bt connection |measurement|
|bt state | |config
|ble transmitter ||diagnostic
|dnd | |config
|device locked, keyguard locked || diagnostic
|device secure, keyguard secure || config
|last reboot | |diagnostic
|last update | |diagnostic
|light sensor, pressure sensor | measurement|
|mobile data, mobile data roaming || config
|wifi ssid, wifi bssid, wifi ip, public ip || diagnostic
|wifi state ||config
|wifi link speed, wifi frequency, wifi signal strength |measurement| diagnostic
|last notification, last removed notification|| none
|active notifications | measurement|
|phone state ||none
|sim1, sim2 || diagnostic
|interactive, doze, power save || diagnostic|
|proximity || diagnostic|
|steps| total_increasing||
|internal/external storage |measurement| diagnostic
|time zone || diagnostic|
|mobile tx, mobile rx, total rx, total tx | total_increasing|diagnostic|

## Screenshots

image

## Link to pull request in Documentation repository

Documentation: home-assistant/companion.home-assistant#

### CODE REVIEW

**Comment by dshokouhi** [comment]
Made a few changes in the last commit:

* Add 2 new DB fields to keep track of core and app registered versions on a per sensor basis, now we will check each sensor and update registration if either the app version has changed or the core version has changed.
* switched back to `getHomeAssistantVersion()` and am now storing the core version with only updating it every 4 hours. So if the update interval has not hit yet we send the cached version then 4 hours later we will send updated version
* SensorReceiver is now responsible for storing and checking the app version locally

**Comment by dshokouhi** [comment]
Ok tested all the latest changes against the depending core PR which was merged recently. During testing totally realized that I submitted the wrong interval, haha. Can confirm re-registration works as expected, I have done testing by manually bumping the core version locally.

File: app/src/main/java/io/homeassistant/companion/android/sensors/SensorReceiver.kt [hunk 1]
```diff
@@ -151,6 +152,16 @@ class SensorReceiver : BroadcastReceiver() {
         val sensorDao = AppDatabase.getInstance(context).sensorDao()
         val enabledRegistrations = mutableListOf<SensorRegistration<Any>>()
 
+        val checkDeviceRegistration = integrationUseCase.getRegistration()
+        if (checkDeviceRegistration.appVersion == null) {
+            Log.w(TAG, "Device not registered, skipping sensor update/registration")
+            return
+        }
+
+        integrationUseCase.setInstalledAppVersion(BuildConfig.VERSION_NAME)
```
Comment (JBassett) [inline app/src/main/java/io/homeassistant/companion/android/sensors/SensorReceiver.kt:161]: Why do we need to store this in the preference?

Comment (dshokouhi) [reply app/src/main/java/io/homeassistant/companion/android/sensors/SensorReceiver.kt:161]: Umm wasn't sure if we would need it in another place. I can remove and check build config directly

**Review by JBassett — APPROVED** [review]

**Comment by CloCkWeRX** [comment]
Interestingly, a number of sensors hadnt re registered since this and the related PR landed, at least for me.
Have removed data from HA, will open specific bugs for specific sensors if it reoccurs.

**Comment by dshokouhi** [comment]
> Interestingly, a number of sensors hadnt re registered since this and the related PR landed, at least for me. Have removed data from HA, will open specific bugs for specific sensors if it reoccurs.

please do not comment on old and merged PRs, instead create a new issue. Make sure to fill out the issue template completely including the requested logs.

**Pull request merged**